{% extends "layout.html" %}

{% block body %}

<div id="banner">
    <div class="container">
        <div class="title">
            <h2>WHAT IS CHIP-IO?</h2>
            <span class="byline">ChIP-IO is a user-friendly web application designed to provide up-to-date, uniformly processed ChIP-Seq data and gene-specific transcription factor regulatory information. </span> 
        </div>
    </div>
</div>

<div id="featured1">
    <div class="container">
        <div class="title">
            <h2>OUR MOTIVATION</h2>
            <span class="byline"></span> </div>
        <p> Chromatin immunoprecipitation followed by sequencing (ChIP-Seq) continues to be a critical informer of protein-DNA binding. As the amount
            of data continues to grow and the needs of researchers continues to become more complex, there is a need for up-to-date, user-friendly tools to serve these
            needs. 
        </p>
        <p>
            With the initiation of projects such as the Human Cell Atlas and the rise of techniques like single-cell RNA-seq,for example, it is clear that the field is moving
            towards more and more complex analysis at narrower and narrower scopes. ChIP-IO was designed with this in mind, allowing for tissue-specific and other 
            metadata-specific analysis to meet increasingly nuanced needs.
        </p>
    </div>
</div>

<div id="featured2">
    <div class="container">
        <div class="title">
            <h2>Technical Details</h2>
            <span class="byline"></span> </div>
        <p> 
            ChIP-IOs underlying analysis process currently consists of two main steps. First is data collection, cleaning, and uniform processing for creation of the core database. 
            Second is the custom-query pipeline for assessing transcription-factor gene regulatory targets. The former is external to this application, but can be used to update 
            the reference datasets periodically. Current processed data can be directly downloaded under the <a style="color: #8fd8f5" href="{{ url_for('downloads') }}">"DOWNLOADS"</a> tab.
            The latter is housed in this application under the <a style="color: #8fd8f5" href="{{ url_for('construct_query') }}">"CONSTRUCT QUERY"</a> tab.
            It handles user-provided queries and actively delivers the results. Details and schematics for both are provided below.
        </p>
    </div>
</div>

<div id="featured1">
    <div class="container">
        <div class="title">
            <h2>DATA COLLECTION</h2>
            <span class="byline"></span> </div>
            <img src={{ url_for('static',filename='images/data_pipeline.png') }}>
            <br>
            <br>
        <p>The out</p>
    </div>
</div>

<div id="featured2">
    <div class="container">
        <div class="title">
            <h2>QUERY PIPELINE</h2>
            <span class="byline"></span> </div>
            <img src={{ url_for('static',filename='images/ChIP-IO_pipeline.png') }}>
            <br>
            <br>
        <p>The out</p>
    </div>
</div>

<div id="featured1">
        <div class="container">
            <div class="title">
                <h2>References</h2>
                <span class="byline"></span> </div>
            <p> Motif Finding</p>
            <p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1539026/</p>
            <p>https://ieeexplore.ieee.org/document/6914361</p>
            <p>https://www.cell.com/cell/pdf/S0092-8674(18)30106-5.pdf</p>
            <p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4121056/</p>
            <p>Gabriel Cuellar-Partida, Fabian A. Buske, Robert C. McLeay, Tom Whitington, William Stafford Noble, and Timothy L. Bailey, "Epigenetic priors for identifying active transcription factor binding sites",
                Bioinformatics 28(1): 56-62, 2012 </p>
            <p>[1] Pique-Regi RP, Degner JF, Pai AA, Gaffney DG, Gilad Y, Pritchard JK. "Accurate inference of transcription factor binding from DNA sequence and chromatin accessibility data", Genome Research. 2011 Mar;21(3):447-55. [Full] [Pubmed]</p>
        </div>
    </div>
{% endblock body %}