{% extends "layout.html" %}

{% block body %}

<div id="banner">
    <div class="container">
        <div class="title">
            <h2>WHAT IS CHIP-IO?</h2>
            <span class="byline"></span> </div>
            <p> ChIP-IO is a user-friendly web application designed to provide up-to-date, uniformly processed ChIP-Seq 
                data and gene-specific transcription factor regulatory information.
            </p>
        </div>
    </div>
</div>

<div id="featured1">
    <div class="container">
        <div class="title">
            <h2>OUR MOTIVATION</h2>
            <span class="byline"></span> </div>
        <p> Chromatin immunoprecipitation followed by sequencing (ChIP-Seq) continues to be a critical informer of protein-DNA binding. As the amount
            of data continues to grow and the needs of researchers continue to become more complex, more can be done to facilitate the use of this critical 
            data source.
        </p>
        <p>
            Furthermore, with the initiation of projects such as the Human Cell Atlas and the rise of techniques like single-cell RNA-seq, it is clear that the 
            field is moving towards more and more complex analysis at narrower and narrower scopes. ChIP-IO was designed with all things considered, hosting
            a large pool of uniformly processed ChIP-seq data and using it for metadata-specific analysis to meet increasingly nuanced demands.
        </p>
    </div>
</div>

<div id="featured2">
    <div class="container">
        <div class="title">
            <h2>Technical Details</h2>
            <span class="byline"></span> </div>
            <p> 
                Full technical details can be found here: (PAPER DOES NOT EXIST YET)
            </p>
            <p> 
                ChIP-IO's underlying analysis currently consists of two main steps. First is data collection, cleansing, and uniform processing for creation of the core database. 
                Second is the custom query pipeline for assessing transcription-factor gene regulatory targets. The former is external to this application, but can be used to update 
                the reference datasets periodically. Current processed data  from the collection step can be directly
                downloaded under the <a style="color: #8fd8f5" href="{{ url_for('downloads') }}">"DOWNLOADS"</a> tab.
                The latter is housed in this application under the <a style="color: #8fd8f5" href="{{ url_for('construct_query') }}">"CONSTRUCT QUERY"</a> tab.
                It handles user-provided queries and actively delivers the results. Details and schematics for both are provided below.
            </p>
    </div>
</div>

<div id="featured1">
    <div class="container">
        <div class="title">
            <h2>DATA COLLECTION</h2>
            <span class="byline"></span> </div>
            <br>
            <br>
            <p> Data was collected from two databases: NCBI's Gene Expression Omnibus (GEO) 
                and the Encyclopedia of DNA Elements(ENCODE). Aligned reads were obtained either by directly or by aligning 
                raw reads using Bowtie2. Study replicates were organized as replicates/controls 
                and then peak calling performed using MACS2 default parameters. Peak parameters
                such as pileup, p-values, etc are in reference to the output of MACS2.
            </p>
            <img src={{ url_for('static',filename='images/Data_Collection_Pipeline.png') }}>
    </div>
</div>

<div id="featured2">
    <div class="container">
        <div class="title">
            <h2>QUERY PIPELINE</h2>
            <span class="byline"></span> </div>
            <br>
            <br>
            <p>All peaks and relevant metadata are collected into a single database. Upon submission of a query, peaks are filtered based on provided peak constrains such as
                p-values and fold-enrichment. Next, valid regulatory regions (promoters, enhancers) are collected depending on other provided constraints and mapped to their
                corresponding genes. Peaks are then annotated to these regions and peak counts bucketed based on the peak's source transcription factor and the regulatory 
                region's gene. These bins consitute the program output. Future work will include epigenetic prior-based motif binding predictions as well as motif discovery 
                based on peak sequences. 
            </p>
            <img src={{ url_for('static',filename='images/ChIP-IO_pipeline.png') }}>
    </div>
</div>

<!-- <div id="featured2">
        <div class="container">
            <div class="title">
                <h2>References</h2>
                <span class="byline"></span> </div>
            <p> Motif Finding</p>
            <p>Barrett T, Wilhite SE, Ledoux P, Evangelista C, Kim IF, Tomashevsky M, Marshall KA, Phillippy KH, Sherman PM, Holko M, Yefanov A, Lee H, Zhang N, Robertson CL, Serova N, Davis S, Soboleva A.
                NCBI GEO: archive for functional genomics data sets--update.
            </p>
            <p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3439153/</p>
            <p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1539026/</p>
            <p>https://ieeexplore.ieee.org/document/6914361</p>
            <p>https://www.cell.com/cell/pdf/S0092-8674(18)30106-5.pdf</p>
            <p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4121056/</p>
            <p>Gabriel Cuellar-Partida, Fabian A. Buske, Robert C. McLeay, Tom Whitington, William Stafford Noble, and Timothy L. Bailey, "Epigenetic priors for identifying active transcription factor binding sites",
                Bioinformatics 28(1): 56-62, 2012 </p>
            <p>[1] Pique-Regi RP, Degner JF, Pai AA, Gaffney DG, Gilad Y, Pritchard JK. "Accurate inference of transcription factor binding from DNA sequence and chromatin accessibility data", Genome Research. 2011 Mar;21(3):447-55. [Full] [Pubmed]</p>
        </div>
    </div> -->
{% endblock body %}